#Data Mining  

#####This notebook will read the pageviews data provided by Google for the year 2015-17 and download additional data for the year 2018-19 

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Import the necessary modules**

In [2]:
import numpy as np
import pandas as pd
import os
import json

# Wikipedia API library
# Allows us to gather page views for different wikipedia articles
!pip install pageviewapi
import pageviewapi

**Set the directory paths**

In [3]:
baseDir = '/content/drive/MyDrive/Time Series Analysis (Mini project)/Time Series Analysis (Mini project)'
datasets = 'Datasets'
datapath1 = 'data_2015-17.csv'
datapath2 = 'json_files'

**Adding the directory containing the custom modules**



In [5]:
import sys
sys.path.append(baseDir + '/utils')

# Custom module to fetch the page views for the given wikipedia article
import wikiAPI 

**Importing the Dataset**

In [6]:
dataframe = pd.read_csv(os.path.join(baseDir, datasets, datapath1))
dataframe.head()

Page  ...  2017-09-10
0            2NE1_zh.wikipedia.org_all-access_spider  ...        38.0
1             2PM_zh.wikipedia.org_all-access_spider  ...        81.0
2              3C_zh.wikipedia.org_all-access_spider  ...         6.0
3         4minute_zh.wikipedia.org_all-access_spider  ...         4.0
4  52_Hz_I_Love_You_zh.wikipedia.org_all-access_s...  ...         7.0

[5 rows x 804 columns]

In [ ]:
dataframe.tail()

**This function takes dataframe and extract the page details.  
The page title is seperated from the domain and access details and returns 2 lists**

In [ ]:
dataset1 = pd.DataFrame(dataset1['Page'], columns = ['Page'])
dataset1

In [ ]:
wikiDf = wikiAPI.extractTitleAndAccessDetails(dataframe)
wikiDf

Page  ... Viewer Type
0                                                    2NE1  ...      spider
1                                                     2PM  ...      spider
2                                                      3C  ...      spider
3                                                 4minute  ...      spider
4                                        52_Hz_I_Love_You  ...      spider
...                                                   ...  ...         ...
145058                    Underworld_(serie_de_películas)  ...      spider
145059                      Resident_Evil:_Capítulo_Final  ...      spider
145060                              Enamorándome_de_Ramón  ...      spider
145061                             Hasta_el_último_hombre  ...      spider
145062  Francisco_el_matemático_(serie_de_televisión_d...  ...      spider

[145063 rows x 4 columns]

**Removing the '.org' from the domain to enable API call**

In [ ]:
wikiDf['Domain'] = wikiDf['Domain'].str.replace('.org', '')
wikiDf

**Checking for missing values in the new dataframe**

In [ ]:
wikiDf.isnull().sum()

**Demonstrating the API call to check the returned JSON data**

In [ ]:
Domain = 'www.mediawiki'
Article = 'Extension:TreeAndMenu'
startDate = '20180101'
endDate = '20191231'
access = 'all-access'
agent = 'all-agents'
granularity='daily'

pageviewapi.per_article(Domain, Article, startDate, endDate, access, agent, granularity)


AttrDict({'items': [{'project': 'zh.wikipedia', 'article': '2NE1', 'granularity': 'daily', 'timestamp': '2018010100', 'access': 'all-access', 'agent': 'spider', 'views': 8}, {'project': 'zh.wikipedia', 'article': '2NE1', 'granularity': 'daily', 'timestamp': '2018010200', 'access': 'all-access', 'agent': 'spider', 'views': 12}, {'project': 'zh.wikipedia', 'article': '2NE1', 'granularity': 'daily', 'timestamp': '2018010300', 'access': 'all-access', 'agent': 'spider', 'views': 25}, {'project': 'zh.wikipedia', 'article': '2NE1', 'granularity': 'daily', 'timestamp': '2018010400', 'access': 'all-access', 'agent': 'spider', 'views': 14}, {'project': 'zh.wikipedia', 'article': '2NE1', 'granularity': 'daily', 'timestamp': '2018010500', 'access': 'all-access', 'agent': 'spider', 'views': 17}, {'project': 'zh.wikipedia', 'article': '2NE1', 'granularity': 'daily', 'timestamp': '2018010600', 'access': 'all-access', 'agent': 'spider', 'views': 25}, {'project': 'zh.wikipedia', 'article': '2NE1', 'gra

**Setting the parameters for the request**

In [ ]:
path = os.path.join(baseDir, datasets, datapath2)
startDate = '20180101'
endDate = '20191231'
granularity = 'daily'

**Call the module and pass the Parameters. Refer the utils folder to examine the fetchwikiData function**

In [ ]:
dataset1['Page'], wikiDf['Page'] = dataset1['Page'].str.replace('/', '%0%0%0'), wikiDf['Page'].str.replace('/', '%0%0%0')
wikiAPI.fetchwikiData(dataset1, wikiDf, path, startDate, endDate, granularity)

**Checking the number of JSON files downloaded. Must be equal to 145062 files in this case.**

In [ ]:
fileCount = os.listdir(path)
fileCount